File: src/test/java/io/github/ajoz/workshop/fp/part_1/exercises/exercise_5/Exercise5Test.java
package io.github.ajoz.workshop.fp.part_1.exercises.exercise_5;

import static io.github.ajoz.workshop.fp.part_1.exercises.exercise_5.Exercise5.composeConsumer;
import static io.github.ajoz.workshop.fp.part_1.exercises.exercise_5.Exercise5.composeSupplier;
import org.junit.Assert;
import org.junit.Test;

public class Exercise5Test {
    @Test
    public void should_compose_function_with_consumer() {
        // given:
        final Consumer1<Integer> consumer =
                integer -> System.out.println(String.format("consumed: %d", integer));

        final Function1<String, Integer> function = Integer::valueOf;

        // when:
        final Consumer1<String> tested = composeConsumer(function, consumer);

        // then:
        tested.accept("42");
    }

    @Test
    public void should_compose_function_with_supplier() {
        // given:
        final Supplier<String> supplier =
                () -> "42";

        final Function1<String, Integer> function = Integer::valueOf;

        // when:
        final Supplier<Integer> tested = composeSupplier(function, supplier);

        // then:
        Assert.assertEquals(new Integer(42), tested.get());
    }
}

File: src/test/kotlin/io/github/ajoz/workshop/fp/kotlin/part_2/solutions/exercise_5/Exercise5Test.kt
@file:Suppress("PackageName")

package io.github.ajoz.workshop.fp.kotlin.part_2.solutions.exercise_5

import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

class Exercise5Test {
    @Test
    fun should_map_ints() {
        // given:
        val ints = Arrays.asList(1, 2, 3)
        val expected = Arrays.asList(2, 3, 4)
        val addOne: (Int) -> Int = { x -> x + 1 }

        // when:
        val actual = mapInts(ints, addOne)

        // then:
        assertEquals(expected, actual)
    }

    @Test
    fun should_map_a_list() {
        // given:
        val ints = Arrays.asList(1, 2, 3)
        val expected = Arrays.asList("1", "2", "3")
        val toString: (Int) -> String = { x -> "$x" }

        // when:
        val actual = map(ints, toString)

        // then:
        assertEquals(expected, actual)
    }

    @Test
    fun test_addOne() {
        // given:
        val ints = Arrays.asList(1, 2, 3)
        val expected = Arrays.asList(2, 3, 4)

        // when:
        val actual = addOne(ints)

        // then:
        assertEquals(expected, actual)
    }

    @Test
    fun test_lengths() {
        // given:
        val strings = Arrays.asList("JUG", "Lodz")
        val expected = Arrays.asList(3, 4)

        // when:
        val actual = lengths(strings)

        // then:
        assertEquals(expected, actual)
    }
}
